### DISCUSSION TITLE
Is it possible to update the certificate profiles using fastlane?

### DISCUSSION THREAD

**Discussion by arunbharath-bfg** [fastlane/fastlane#22071] [Q&A]
I'm trying refresh my profile using Fastlane. But I'm getting error at profile.update! as function not found. I tried google to figure out the method for update operation, but I couldn't find anything. any help would be appreciated. I'm trying to fetch certificates from apple store and renew the expired one

```ruby
def profile_refresh(team_name, git)
  verify_git_branch(team_name, git)
  UI.message "*************************"
  UI.message "Beginning profile refresh"
  UI.message "*************************"

  certificates = Spaceship::ConnectAPI::Certificate.all
  UI.message "profile refresh Certificate list"
  devices = Spaceship::ConnectAPI::Device.all
  teams = ["XXX",[YYY]]

  team_devices = devices.select do |device|
    team_code = device.name.split(" ").first
    teams.include? team_code
  end
  Spaceship::ConnectAPI::Profile.all.each do |profile|
    UI.message "status #{profile.profile_state}"
    UI.message "status #{profile.profile_type}" 
    if profile.platform == "tvOS"
      UI.important "#{profile.name} is a tvOS profile, skipping."
    else
      UI.message "Updating #{profile}"
      profile.certificates = certificates
      profile.devices = team_devices
      profile.update!
    end
  end
``` 

```groovy
18:36:38  ERROR [2024-06-05 22:36:38.70]: Error in your Fastfile at line 140
18:36:38  ERROR [2024-06-05 22:36:38.71]:     138:	      profile.certificates = certificates
18:36:38  ERROR [2024-06-05 22:36:38.71]:     139:	      profile.devices = team_devices
18:36:38  ERROR [2024-06-05 22:36:38.71]:  => 140:	      profile.update!
18:36:38  ERROR [2024-06-05 22:36:38.71]:     141:	    end
18:36:38  ERROR [2024-06-05 22:36:38.71]:     142:	  end
18:36:38  ERROR [2024-06-05 22:36:38.71]: fastlane finished with errors
18:36:38  
18:36:38  Looking for related GitHub issues on fastlane/fastlane...
18:36:38  Search query: undefined method `update!' for an instance of Spaceship::ConnectAPI::Profile
18:36:38  
18:36:38  URL: https://api.github.com/search/issues?q=undefined%20method%20%60update%21%27%20for%20an%20instance%20of%20Spaceship%3A%3AConnectAPI%3A%3AProfile+repo:fastlane/fastlane
18:36:38  ➡️  [match] add caching layer to significantly improve performance by up to 100x
18:36:38      https://github.com/fastlane/fastlane/pull/21694 [closed] 16 💬
18:36:38      04 Jan 2024
18:36:38  
18:36:38  🔗  You can ⌘ + double-click on links to open them directly in your browser.
18:36:38  Fastfile:140:in `block in profile_refresh': [!] undefined method `update!' for an instance of Spaceship::ConnectAPI::Profile (NoMethodError)
```

**Comment by florence0239** [comment]
> I'm trying refresh my profile using Fastlane. But I'm getting error at profile.update! as function not found. I tried google to figure out the method for update operation, but I couldn't find anything. any help would be appreciated. I'm trying to fetch certificates from apple store and renew the expired one
> 
> ```ruby
> def profile_refresh(team_name, git)
>   verify_git_branch(team_name, git)
>   UI.message "*************************"
>   UI.message "Beginning profile refresh"
>   UI.message "*************************"
> 
>   certificates = Spaceship::ConnectAPI::Certificate.all
>   UI.message "profile refresh Certificate list"
>   devices = Spaceship::ConnectAPI::Device.all
>   teams = ["XXX",[YYY]]
> 
>   team_devices = devices.select do |device|
>     team_code = device.name.split(" ").first
>     teams.include? team_code
>   end
>   Spaceship::ConnectAPI::Profile.all.each do |profile|
>     UI.message "status #{profile.profile_state}"
>     UI.message "status #{profile.profile_type}" 
>     if profile.platform == "tvOS"
>       UI.important "#{profile.name} is a tvOS profile, skipping."
>     else
>       UI.message "Updating #{profile}"
>       profile.certificates = certificates
>       profile.devices = team_devices
>       profile.update!
>     end
>   end
> ```
> 
> ```groovy
> 18:36:38  ERROR [2024-06-05 22:36:38.70]: Error in your Fastfile at line 140
> 18:36:38  ERROR [2024-06-05 22:36:38.71]:     138:	      profile.certificates = certificates
> 18:36:38  ERROR [2024-06-05 22:36:38.71]:     139:	      profile.devices = team_devices
> 18:36:38  ERROR [2024-06-05 22:36:38.71]:  => 140:	      profile.update!
> 18:36:38  ERROR [2024-06-05 22:36:38.71]:     141:	    end
> 18:36:38  ERROR [2024-06-05 22:36:38.71]:     142:	  end
> 18:36:38  ERROR [2024-06-05 22:36:38.71]: fastlane finished with errors
> 18:36:38  
> 18:36:38  Looking for related GitHub issues on fastlane/fastlane...
> 18:36:38  Search query: undefined method `update!' for an instance of Spaceship::ConnectAPI::Profile
> 18:36:38  
> 18:36:38  URL: https://api.github.com/search/issues?q=undefined%20method%20%60update%21%27%20for%20an%20instance%20of%20Spaceship%3A%3AConnectAPI%3A%3AProfile+repo:fastlane/fastlane
> 18:36:38  ➡️  [match] add caching layer to significantly improve performance by up to 100x
> 18:36:38      https://github.com/fastlane/fastlane/pull/21694 [closed] 16 💬
> 18:36:38      04 Jan 2024
> 18:36:38  
> 18:36:38  🔗  You can ⌘ + double-click on links to open them directly in your browser.
> 18:36:38  Fastfile:140:in `block in profile_refresh': [!] undefined method `update!' for an instance of Spaceship::ConnectAPI::Profile (NoMethodError)
> ```

Hello, [healthcare gov marketplace](https://www.health-insurancemarket.com)

It appears that the method update! does not exist for Spaceship::ConnectAPI::Profile. Instead, you need to use the correct method to update the profile, which is typically handled via a different API call or update pattern. Below is an updated version of your script to correctly update the profiles.

Here is how you can do it:

Fetch the existing profile.
Update the profile using Spaceship::ConnectAPI::Profile.update_profile method.
Here’s how you can modify your script:
` `require 'fastlane'
require 'spaceship'

def profile_refresh(team_name, git)
  verify_git_branch(team_name, git)
  UI.message "*************************"
  UI.message "Beginning profile refresh"
  UI.message "*************************"

  certificates = Spaceship::ConnectAPI::Certificate.all
  UI.message "profile refresh Certificate list"
  devices = Spaceship::ConnectAPI::Device.all
  teams = ["XXX", "YYY"]

  team_devices = devices.select do |device|
    team_code = device.name.split(" ").first
    teams.include? team_code
  end

  Spaceship::ConnectAPI::Profile.all.each do |profile|
    UI.message "status #{profile.profile_state}"
    UI.message "status #{profile.profile_type}"
    if profile.platform == "tvOS"
      UI.important "#{profile.name} is a tvOS profile, skipping."
    else
      UI.message "Updating #{profile.name}"

      # Update profile with new certificates and devices
      Spaceship::ConnectAPI::Profile.update_profile(
        profile_id: profile.id,
        bundle_id: profile.bundle_id,
        name: profile.name,
        certificates: certificates,
        devices: team_devices,
        profile_type: profile.profile_type
      )

      UI.message "#{profile.name} updated successfully."
    end
  end
end``

Explanation
Remove profile.update!: This method does not exist for Spaceship::ConnectAPI::Profile.
Use Spaceship::ConnectAPI::Profile.update_profile: This method correctly updates the profile with new certificates and devices.
Important Notes
Ensure that you have the latest version of the fastlane and spaceship gems.
Handle the authentication and initialization of Spaceship before calling profile_refresh.
This should correctly handle the updating of your profiles with the new certificates and devices.

Hope this will help you.

Best Regard,
florence0239
